Sources/NIOWebSocket/WebSocketUpgrader.swift

```diff
@@ -0,0 +1,126 @@
+        guard fields.count == 1 else {
+            throw NIOWebSocketUpgradeError.invalidUpgradeHeader
+        }
+        return fields.first!
+    }
+}
+
+/// A `HTTPProtocolUpgrader` that knows how to do the WebSocket upgrade dance.
+///
+/// Users may frequently want to offer multiple websocket endpoints on the same port. For this
+/// reason, this `WebSocketUpgrader` only knows how to do the required parts of the upgrade and to
+/// complete the handshake. Users are expected to provide a callback that examines the HTTP headers
+/// (including the path) and determines whether this is a websocket upgrade request that is acceptable
+/// to them.
+///
+/// This upgrader assumes that the `HTTPServerUpgradeHandler` will appropriately mutate the pipeline to
+/// remove the HTTP `ChannelHandler`s.
+public final class WebSocketUpgrader: HTTPProtocolUpgrader {
+    /// RFC 6455 specs this as the required entry in the Upgrade header.
+    public let supportedProtocol: String = "websocket"
+
+    /// We deliberately do not actually set any required headers here, because the websocket
+    /// spec annoyingly does not actually force the client to send these in the Upgrade header,
+    /// which NIO requires. We check for these manually.
+    public let requiredUpgradeHeaders: [String] = []
+
+    private let shouldUpgrade: (HTTPRequestHead) -> HTTPHeaders?
+    private let upgradePipelineHandler: (Channel, HTTPRequestHead) -> EventLoopFuture<Void>
+
+    /// Create a new `WebSocketUpgrader`.
+    ///
+    /// - parameters:
+    ///     - shouldUpgrade: A callback that determines whether the websocket request should be
+    ///         upgraded. This callback is responsible for creating a `HTTPHeaders` object with
+    ///         any headers that it needs on the response *except for* the `Upgrade`, `Connection`,
+    ///         and `Sec-WebSocket-Accept` headers, which this upgrader will handle. Should return
+    ///         `nil` if the upgrade should be refused.
+    ///     - upgradePipelineHandler: A function that will be called once the upgrade response is
+    ///         flushed, and that is expected to mutate the `Channel` appropriately to handle the
+    ///         websocket protocol. This only needs to add the user handlers: the
+    ///         `WebSocketFrameEncoder` and `WebSocketFrameDecoder` will have been added to the
+    ///         pipeline automatically.
+    public init(shouldUpgrade: @escaping (HTTPRequestHead) -> HTTPHeaders?,
+                upgradePipelineHandler: @escaping (Channel, HTTPRequestHead) -> EventLoopFuture<Void>) {
+        self.shouldUpgrade = shouldUpgrade
+        self.upgradePipelineHandler = upgradePipelineHandler
+    }
+
+    public func buildUpgradeResponse(upgradeRequest: HTTPRequestHead, initialResponseHeaders: HTTPHeaders) throws -> HTTPHeaders {
+        let key = try upgradeRequest.headers.getNonListHeader("Sec-WebSocket-Key").trimmingCharacters(in: .whitespaces)
```

Comment:
this would be nicer as a subscript with label, no?

    try upgradeRequest.headers[nonListHeader: "Sec-WebSocket-Key"]

or

    try upgradeRequests.headers.nonListHeaders["Sec-WebSocket-Key"]

?

Reply:
Swift subscripts cannot throw.

Reply:
why not return `String?` and throw here? Might be better anyway as that extension might prove useful in other situations?